### DISCUSSION TITLE
how to provide system properties?

### DISCUSSION THREAD

**Discussion by axaluss** [com-lihaoyi/mill#1380] [Q&A] [answered]
I'ld like to provide system properties. how can i do this?

Logically i'd like to do: 
```bash
mill -Dlogback.configurationFile=${CI_PROJECT_DIR}/bin/logback-ci.xml -j 0 __.test
```

**Comment by hmf** [comment]
If you are in Linux, try something like:

JAVA_OPTS="-DsocksProxyHost=127.0.0.1 -DsocksProxyPort=1080 -DsocksProxyVersion=5" mill -i core.console

ss indicate [here](https://gitter.im/lihaoyi/mill?at=5da124f180e62056e4eb63b1).

HTHs

**Reply by axaluss** [reply]
Thank you! I searched so hard but couldn't find it. I think this belongs on the docs.

**Reply by lefou** [reply]
This only works when you run mill with `-i` or `--no-server`. See my other answer.

**Comment by lefou** [comment, marked as answer]
Use `-D` or `--define` option:

```
mill -D logback.configurationFile=${CI_PROJECT_DIR}/bin/logback-ci.xml -j 0 __.test
```

Notice the space between the option and its argument.

**Reply by axaluss** [reply]
Yeah that one works locally but not on the remote CI machine.
It still insists on using `logback-test.xml` instead
` mill -Dlogback.configurationFile=${CI_PROJECT_DIR}/bin/logback-ci.xml -j 0 __.test    ` 
is running fine locally picking up the `logback-ci.xml`.

On the remote it's running
```
/usr/local/bin/mill -D ivy.home=/builds/pipeline/my-service/.cache \
 -D logback.configurationFile=/builds/pipeline/my-service/bin/logback-ci.xml \
 --home /builds/pipeline/my-service/.mill/ -j 0 --color \
 false --disable-ticker --silent -j 0 -D logback.configurationFile=/builds/pipeline/my-service/bin/logback-ci.xml __.test
  
.....  

|-INFO in ch.qos.logback.classic.LoggerContext[default] - Found resource [logback-test.xml] at [file:/builds/pipeline/my-service/my_service/test/resources/logback-test.xml]
```
🤔

**Reply by lefou** [reply]
Please add a `--debug` and inspect the head of the output. Look for a line starting with: "Using explicit system properties:"

**Reply by axaluss** [reply]
i'm not sure what happened, but it's working now. Thank you!

**Reply by axaluss** [reply]
@lefou Ok this is not working reliably.  at least with version 0.9.8 `-D` value `logback.configurationFile`  is printed in the `--debug` output correctly, but It's not applied reliably. `-i` flag is not helping either.
No idea why this sometimes works and sometimes not...

**Reply by lefou** [reply]
One thing I could imagine is timing. Maybe, you access the system property very early, e.g. in some static initialization context. In such a case, it is probably read before we set it.
